packages/@webex/plugin-meetings/src/meeting/index.ts

```diff
@@ -4316,34 +4316,71 @@ export default class Meeting extends StatelessWebexPlugin {
         new ParameterError('joinWithMedia() can only be used with allowMediaInLobby set to true')
       );
     }
+    this.allowMediaInLobby = true;
 
     LoggerProxy.logger.info('Meeting:index#joinWithMedia called');
 
-    return this.join(joinOptions)
-      .then((joinResponse) =>
-        this.addMedia(mediaOptions).then((mediaResponse) => ({
-          join: joinResponse,
-          media: mediaResponse,
-        }))
-      )
-      .catch((error) => {
-        LoggerProxy.logger.error('Meeting:index#joinWithMedia --> ', error);
+    let joined = false;
 
-        Metrics.sendBehavioralMetric(
-          BEHAVIORAL_METRICS.JOIN_WITH_MEDIA_FAILURE,
-          {
-            correlation_id: this.correlationId,
-            locus_id: this.locusUrl.split('/').pop(),
-            reason: error.message,
-            stack: error.stack,
-          },
-          {
-            type: error.name,
-          }
-        );
+    try {
+      let turnServerInfo;
+      let turnDiscoverySkippedReason;
 
-        return Promise.reject(error);
-      });
+      // @ts-ignore
+      joinOptions.reachability = await this.webex.meetings.reachability.getReachabilityResults();
+      const turnDiscoveryRequest = await this.roap.generateTurnDiscoveryRequestMessage(this, true);
+
+      turnDiscoverySkippedReason = turnDiscoveryRequest.skippedReason;
+      joinOptions.roapMessage = turnDiscoveryRequest.roapMessage;
+
+      const joinResponse = await this.join(joinOptions);
+
+      joined = true;
+
+      if (joinOptions.roapMessage) {
+        ({turnServerInfo, turnDiscoverySkippedReason} =
+          await this.roap.handleTurnDiscoveryHttpResponse(this, joinResponse));
+
+        this.turnDiscoverySkippedReason = turnDiscoverySkippedReason;
+        this.turnServerUsed = !!turnServerInfo;
+      }
+
+      const mediaResponse = await this.addMedia(mediaOptions, turnServerInfo);
+
+      return {
+        join: joinResponse,
+        media: mediaResponse,
+      };
+    } catch (error) {
+      LoggerProxy.logger.error('Meeting:index#joinWithMedia --> ', error);
+
+      let leaveError;
+
+      if (joined) {
+        try {
+          await this.leave({resourceId: joinOptions?.resourceId, reason: 'joinWithMedia failure'});
```

Comment:
some other reasons that are passed through to ```this.leave(``` are constants. Do we want this to also be a constant somewhere for consistency?

Reply:
if you follow the existing code closely you'll see that there are constants from different enums being passed in as reason - for example sometimes they are from CALL_REMOVED_REASON, sometimes they are from MEETING_REMOVED_REASON, so it just shows that using constants doesn't guarantee that there will be no mess. The `reason` parameter needs to be typed properly first and then all the calls to leave() need to be fixed and this would be big enough change to be a separate PR on its own, so I don't want to do this as part of this work.